core/services/relay/evm/headreporter/head_reporter.go

```diff
@@ -0,0 +1,149 @@
+package headreporter
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/sqlutil"
+	"github.com/smartcontractkit/chainlink/v2/core/services/telemetry"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+	"github.com/smartcontractkit/chainlink-common/pkg/utils/mailbox"
+
+	evmtypes "github.com/smartcontractkit/chainlink/v2/core/chains/evm/types"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/legacyevm"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+)
+
+type (
+	HeadReporter interface {
+		ReportNewHead(ctx context.Context, head *evmtypes.Head) error
+		ReportPeriodic(ctx context.Context) error
+	}
+
+	HeadReporterService struct {
+		services.StateMachine
+		ds             sqlutil.DataSource
+		chains         legacyevm.LegacyChainContainer
+		lggr           logger.Logger
+		newHeads       *mailbox.Mailbox[*evmtypes.Head]
+		chStop         services.StopChan
+		wgDone         sync.WaitGroup
+		reportPeriod   time.Duration
+		reporters      []HeadReporter
+		unsubscribeFns []func()
+	}
+)
+
+func NewHeadReporterService(ds sqlutil.DataSource, chainContainer legacyevm.LegacyChainContainer, lggr logger.Logger, monitoringEndpointGen telemetry.MonitoringEndpointGenerator, opts ...interface{}) *HeadReporterService {
+	reporters := make([]HeadReporter, 1, 2)
+	reporters[0] = NewPrometheusReporter(ds, chainContainer, lggr, opts)
+	telemetryEnabled := false
+	for _, chain := range chainContainer.Slice() {
+		if chain.Config().EVM().HeadTracker().HeadTelemetryEnabled() {
+			telemetryEnabled = true
+			break
+		}
+	}
+	if telemetryEnabled {
+		reporters = append(reporters, NewTelemetryReporter(chainContainer, monitoringEndpointGen))
+	}
+	return NewHeadReporterServiceWithReporters(ds, chainContainer, lggr, reporters, opts)
+}
+
+func NewHeadReporterServiceWithReporters(ds sqlutil.DataSource, chainContainer legacyevm.LegacyChainContainer, lggr logger.Logger, reporters []HeadReporter, opts ...interface{}) *HeadReporterService {
+	reportPeriod := 30 * time.Second
+	for _, opt := range opts {
+		switch v := opt.(type) {
+		case time.Duration:
+			reportPeriod = v
+		default:
+			lggr.Debugf("Unknown opt type '%T' passed to HeadReporterService", v)
```

Comment:
We must either return an error, or provide an API which does not allow these kind of mistakes. If optional params are truly necessary, then we need to use proper functional options or equivalent. However, it looks like it would be overkill in this case.

Reply:
applied from your pr